receiver_test.go

```diff
@@ -47,7 +45,289 @@ func TestReceive_ModeFirst(t *testing.T) {
+	require.Equal(t, errdesc, de.RemoteError.Description)
+	require.NoError(t, client.Close())
+	_, err = r.Receive(context.Background())
+	if !errors.As(err, &de) {
+		t.Fatalf("unexpected error type %T", err)
+	}
+}
+
+func TestReceiveInvalidMessage(t *testing.T) {
+	const linkHandle = 0
+	deliveryID := uint32(1)
+	responder := func(req frames.FrameBody) ([]byte, error) {
+		b, err := receiverFrameHandler(ModeFirst)(req)
+		if b != nil || err != nil {
+			return b, err
+		}
+		switch req.(type) {
+		case *frames.PerformFlow:
+			return nil, nil
+		case *frames.PerformDisposition:
+			return mocks.PerformDisposition(encoding.RoleSender, 0, deliveryID, nil, &encoding.StateAccepted{})
+		default:
+			return nil, fmt.Errorf("unhandled frame %T", req)
+		}
+	}
+	conn := mocks.NewNetConn(responder)
+	client, err := New(conn)
+	assert.NoError(t, err)
+	session, err := client.NewSession()
+	assert.NoError(t, err)
+	r, err := session.NewReceiver()
+	assert.NoError(t, err)
+
+	msgChan := make(chan *Message)
+	errChan := make(chan error)
+	go func() {
+		msg, err := r.Receive(context.Background())
+		msgChan <- msg
+		errChan <- err
+	}()
+
+	// missing DeliveryID
+	fr, err := mocks.EncodeFrame(mocks.FrameAMQP, 0, &frames.PerformTransfer{
+		Handle: linkHandle,
+	})
+	assert.NoError(t, err)
+	conn.SendFrame(fr)
+
+	assert.Nil(t, <-msgChan)
+	var amqpErr *Error
+	if err = <-errChan; !errors.As(err, &amqpErr) {
+		t.Fatalf("unexpected error %v", err)
+	}
+	assert.Equal(t, ErrorNotAllowed, amqpErr.Condition)
+
+	_, err = r.Receive(context.Background())
+	if !errors.As(err, &amqpErr) {
+		t.Fatalf("unexpected error %v", err)
+	}
+
+	// missing MessageFormat
+	r, err = session.NewReceiver()
+	assert.NoError(t, err)
+	go func() {
+		msg, err := r.Receive(context.Background())
+		msgChan <- msg
+		errChan <- err
+	}()
+	fr, err = mocks.EncodeFrame(mocks.FrameAMQP, 0, &frames.PerformTransfer{
+		DeliveryID: &deliveryID,
+		Handle:     linkHandle,
+	})
+	assert.NoError(t, err)
+	conn.SendFrame(fr)
+
+	assert.Nil(t, <-msgChan)
+	if err = <-errChan; !errors.As(err, &amqpErr) {
+		t.Fatalf("unexpected error %v", err)
+	}
+	assert.Equal(t, ErrorNotAllowed, amqpErr.Condition)
+
+	_, err = r.Receive(context.Background())
+	if !errors.As(err, &amqpErr) {
+		t.Fatalf("unexpected error %v", err)
+	}
+
+	// missing delivery tag
+	format := uint32(0)
+	r, err = session.NewReceiver()
+	assert.NoError(t, err)
+	go func() {
+		msg, err := r.Receive(context.Background())
+		msgChan <- msg
+		errChan <- err
+	}()
+	fr, err = mocks.EncodeFrame(mocks.FrameAMQP, 0, &frames.PerformTransfer{
+		DeliveryID:    &deliveryID,
+		Handle:        linkHandle,
+		MessageFormat: &format,
+	})
+	assert.NoError(t, err)
+	conn.SendFrame(fr)
+
+	assert.Nil(t, <-msgChan)
+	if err = <-errChan; !errors.As(err, &amqpErr) {
+		t.Fatalf("unexpected error %v", err)
+	}
+	assert.Equal(t, ErrorNotAllowed, amqpErr.Condition)
+
+	_, err = r.Receive(context.Background())
+	if !errors.As(err, &amqpErr) {
```

Comment:
One thing that's a bit odd here (unless I'm reading the test incorrectly) is that the ErrorNotAllowed is not a violation on our side, it's a violation from the sender.

So it seems like the NotAllowed error should be sent to the service (like you do) but on our side perhaps it's an internal error? Or will this just result in a detach error anyways?

Reply:
It's not a `DetachError` in this case, just an `Error` since we initiated the detach (we only return `DetachError` when the peer initiates the detach).

Reply:
This might be my misunderstanding, although I did fire up the debugger just now to see if I got what this was doing.

Internally we do have an error, but the error is not technically on our side - the problem is that the sender did something incorrect. So reporting an ErrorNotAllowed to the sender makes total sense. But we're also returning our internally generated error to the caller of Receive, where ErrorNotAllowed doesn't quite make sense (ie, the receiver didn't do something illegal, the sender did).

So I was thinking the condition should probably morph to 'ErrorInternalError', or similar, since that's the representation of what's happening for the the client in this situation.

What do you think?

Reply:
While I think there's value in returning what we send to the peer, I do see your point.  If you feel strongly about this can you open an issue to track it?  Changing the behavior here is non-trivial as `l.err` is what we send to the peer and what we return to the caller, so splitting the behavior should be done separately.

Reply:
Will do. I think in this case we wouldn't throw the error away, we'd just wrap it in a suitable error, like it's an inner exception.

Reply:
Filed: https://github.com/Azure/go-amqp/issues/109